src/psfmachine/machine.py

```diff
@@ -882,56 +902,108 @@ def build_time_model(self, plot=False, downsample=False):
-        k &= (
-            (flux_binned_raw > 10).any(axis=0)[None, :]
-            * np.ones(flux_binned_raw.shape, bool)
-        ).ravel()
-        # No huge variability
-        k &= (
-            (np.abs(flux_binned - 1) < 1).all(axis=0)[None, :]
-            * np.ones(flux_binned.shape, bool)
-        ).ravel()
-        # No nans
-        k &= np.isfinite(flux_binned.ravel()) & np.isfinite(flux_err_binned.ravel())
         prior_sigma = np.ones(A3.shape[1]) * 10
         prior_mu = np.zeros(A3.shape[1])
 
-        for count in [0, 1, 2]:
-            sigma_w_inv = A3[k].T.dot(
-                (A3.multiply(1 / flux_err_binned.ravel()[:, None] ** 2)).tocsr()[k]
+        # fit the time model for each segment
+        # we find the splits in data
+        if split_models:
+            dts = np.diff(self.time)
+            splits = np.append(
+                np.append(0, np.where(dts > 5 * np.median(dts))[0] + 1), len(self.time)
             )
-            sigma_w_inv += np.diag(1 / prior_sigma ** 2)
-            # Fit the flux - 1
-            B = A3[k].T.dot(
-                ((flux_binned.ravel() - 1) / flux_err_binned.ravel() ** 2)[k]
+        else:
+            splits = np.array([0, len(self.time)])
+
+        seg_time_model_w, pix_mask_k = [], []
+        # we iterate over segements
+        for bk in range(len(splits) - 1):
+            # find the right mask that select the binned times andd flux to fit the
+            # time model
+            seg_mask = (time_binned[:, 0] >= time_original[splits[bk]]) & (
+                time_binned[:, 0] < time_original[splits[bk + 1] - 1]
             )
-            B += prior_mu / (prior_sigma ** 2)
-            time_model_w = np.linalg.solve(sigma_w_inv, B)
-            res = flux_binned - A3.dot(time_model_w).reshape(flux_binned.shape)
-            res = np.ma.masked_array(res, (~k).reshape(flux_binned.shape))
-            bad_targets = sigma_clip(res, sigma=5).mask
-            bad_targets = (
-                np.ones(flux_binned.shape, bool) & bad_targets.any(axis=0)
+            print(seg_mask)
+            # need to rebuild the A3 DM to use the rigth times/poscorrs
+            A2 = sparse.vstack([A_c] * time_binned[seg_mask].shape[0], format="csr")
+            if hasattr(self, "pos_corr1") and self.time_corrector in [
+                "pos_corr",
+                "centroid",
+            ]:
+                A3 = _combine_A(
+                    A2, poscorr=[poscorr1_binned[seg_mask], poscorr2_binned[seg_mask]]
+                )
+            else:
+                A3 = _combine_A(A2, time=time_binned[seg_mask])
+            print(A3.shape)
+
+            # No saturated pixels
+            k = (
+                (flux_binned_raw < 1.4e5).any(axis=0)[None, :]
+                * np.ones(flux_binned_raw[seg_mask].shape, bool)
+            ).ravel()
+            # No faint pixels
+            k &= (
+                (flux_binned_raw[seg_mask] > 10).any(axis=0)[None, :]
```

Comment:
I bet we could make that more aggressive. Shouldn't it be `all` and shouldn't the value be higher? 10 counts is hardly any, maybe all over 100 counts?

```python
k &= (
    (flux_binned_raw[seg_mask] > 100).all(axis=0)[None, :] 
    * np.ones(flux_binned_raw[seg_mask].shape, bool)
    ).ravel()
```

Reply:
Agree with the count limit. 
Yes, should be `all` to clip pixels that are always within the limits